### ISSUE TITLE
Suggestion: route format character for ShortGuids

### ISSUE THREAD

**Issue by rmunn** [giraffe-fsharp/Giraffe#282]
[ShortGuids](https://madskristensen.net/blog/A-shorter-and-URL-friendly-GUID) are basically GUIDs run through a base64url encoding step (i.e., base64 with `-` and `_` replacing `+` and `/`) and with the final `=` padding characters trimmed because the length is fixed: a ShortGuid will always encode to 22 characters (which will 22 bytes in UTF-8 since they're all from the ASCII range). Youtube uses a similar system for their video IDs, though in Youtube's case the IDs are 64-bit numbers so they encode to 11 characters rather than 22.

I propose adding a regex for ShortGuids in URLs, which will decode to a Guid object (no need for an external library). The regex would be simply `([-_0-9A-Za-z]{22})`. You can generate some sample ShortGuids (such as `IxC0lRDiNkqa_xsaUWbT-Q`) to test against by going to https://www.guidgenerator.com/online-guid-generator.aspx and checking the **Base64 encode** and **RFC 7515** boxes.

The more difficult question is which format character to use for ShortGuids in a route format string. Possibilities:

* `%o` (lowercase `O`). Pro: like GUIDs (`%O`) but "smaller". Con: potential for confusion with `%O` in many fonts, and it already means "octal" numbers.
* `%g` for "Guid", lowercase because it's a "small" Guid. Potential for confusion since `%G` does *not* mean "normal" Guid. If `%G` was also assigned the meaning of normal Guid, now there are two different format markers for Guid (`%G` and `%O`), again leading to confusion.
* `%s` for "short" and `%i` for "id" are already taken for strings and ints, so they're out.
* `%x` for "short" (or "xhort", I guess) is a real stretch, and it . I don't like this one much, though see below.

Of these options, I think `%o` (lowercase `O`) is the least bad. It's a legal character in `PrintfFormat` strings, and I doubt that parsing octal numbers in a URL is a desirable feature for most people so losing that option isn't much of a loss.

Another possibility, that I think is less good but should be mentioned, would be to use `%X` for the 22-character ShortGuid format type, and `%x` for the 11-character UID based on a 64-bit number (e.g., Youtube video IDs). The drawback of this one is that this overloads the meaning of "hex", and there might be some people who would want to have hexadecimal numbers in URLs.

Which brings up one more possibility: using `%G` (capital G) for these URL-formatted GUIDs, and `%g` (lowercase G) for the corresponding 11-character URL-formatted ID. The `%G` and `%g` formats normally specify floating-point values, and overriding them to mean ShortGuids (of both the 22- and 11-byte variety) would not be a loss since floating-point values in URLs are not useful anyway.

~~So my recommendation is:~~ **(UPDATE: I no longer recommend this; see comments below)**

* If 11-character IDs are not desirable, use `%o`.
* If both 11- and 22-character IDs are desirable, use `%G` for a 22-character base64url-encoded GUID that parses to a normal System.Guid object, and `%g` for an 11-character value that parses to an int64.

I would also recommend providing helper functions to convert a System.Guid to a 22-character `%o`/`%G` representation, and (if the 11-character version is chosen) to convert an int64 to an 11-character `%g` representation.

I welcome comments from anyone else interested in this idea, especially on what format character(s) to use for ShortGuids and whether the 11-character, int64 version would be useful as well.

P.S. Another alternative, which I just realized, would be to allow the `%O` format specifier to parse 22-character GUIDs as well. Since the first 22 characters of a 32-character GUID-without-hyphens would be a valid ShortGuid, the 22-character version would need to go on the end of the regex:

```fsharp
let guidFormatStr =
    "([0-9A-Fa-f]{8}\-[0-9A-Fa-f]{4}\-[0-9A-Fa-f]{4}\-[0-9A-Fa-f]{4}\-[0-9A-Fa-f]{12}|[0-9A-Fa-f]{32}|[-_0-9A-Za-z]{22})"
```

That way when `IxC0lRDiNkqa_xsaUWbT-Q` fails the `[0-9A-Fa-f]{32}` regex, it will go on to match `[-_0-9A-Za-z]{22}` instead, but `6d96ec4f084b25602d50a3770e41ae54` would be matched by `[0-9A-Fa-f]{32}` and so the `[-_0-9A-Za-z]{22}` pattern would never be examined.

**Comment by rmunn** [comment]
After thinking it over, the alternative mentioned in my P.S. is the one I'd recommend: extend the formats accepted by `%O` to include the 22-character base64url representation of a Guid, and parse that to a System.Guid value rather than to a ShortGuid value. One of the benefits of this approach is that in code like [Saturn's `Controller.fs`](https://github.com/SaturnFramework/Saturn/blob/master/src/Saturn/Controller.fs), which automatically creates route format strings from a `controller { ... }` CE, the code won't need to change. It currently says "If the data type of your keys is Guid, put a `/%O/` segment in the route string". If Guids could be parsed by either `%O` or `%o`, with `%o` being the format code for a shorter encoding, then Saturn's code would have to change to figure out the correct code to output for Guid typed keys. But if the regex for `%O` is simply expanded to accept the 22-character format, Saturn's code (and any similar code that others may have already written) won't be affected.

**Comment by dustinmoris** [comment]
Thanks for suggesting this feature, which I think is great! I agree with your recommendation of making the current Guid formatter `'O'` also responsible for short Guids. From a user's point of view they just care about the model to which the route binds and if the model holds a `Guid` then any data format which can be resolved into a `Guid` should essentially be valid.

If you want to work on this feature and provide a PR then let me know, otherwise I can also try to pick this up for the next release!

**Comment by dustinmoris** [comment]
@rmunn I have added this feature now based on our discussion here. I have also updated tests and the documentation. Please have a look at the referenced commit and let me know if you see any improvements that can be made.

The only thing I had to do differently that what we discussed here is to use the `%u` format character for the `uint64` YoutTube IDs, because `%u` is the default char for unsigned ints and also because `%o` is already reserved for octal values.

**Comment by rmunn** [comment]
Looks good to me! I found one typo, but otherwise I don't see anything that was missed.

I was initially a little bit bothered by the difference in handling between short GUIDs and short IDs: GUIDs can handle either their "long" form (8-4-4-4-12) or their short form (22 base64url chars), but short IDs only allow the 11-char string form. But after a bit of thought, I realized that there's no getting around it. If you allowed `%u` to match normal uint64s as well, then what about the URL `/foo/12345678901`? That's a valid uint64 *and* a valid shortID, so which one is intended? So the `%u` code has to only allow one form, and the shortID form is going to be more useful than the "normal" uint64 representation.

Thank you for implementing this!
